nx/lib/nx/lin_alg.ex

```diff
@@ -1117,38 +1117,59 @@ defmodule Nx.LinAlg do
         48.0
       >
   """
-  def determinant(tensor)
+  defn determinant(tensor) do
+    Nx.Defn.Kernel.assert_shape_pattern(tensor, {n, n})
+
+    {n, _} = Nx.shape(tensor)
+
+    cond do
+      Nx.equal(n, 2) ->
+        determinant_2by2(tensor)
+
+      Nx.equal(n, 3) ->
+        determinant_3by3(tensor)
+
+      :otherwise ->
+        determinant_NbyN(tensor)
+    end
```

Comment:
I would do this inside a transform because we don't want to the comparison and the branching inside `defn` (i.e. inside the gpu). Something like:

```suggestion
    transform(n, fn
      2 -> determinant_2by2(tensor)
      3 -> determinant_3by3(tensor)
      _ -> determinant_NbyN(tensor)
    end)
```